pulp_service/pulp_service/tasking/new_worker.py

```diff
@@ -541,59 +541,65 @@ def fetch_task(self):
+                # 24 hours = 86400 seconds
+                acquired = self.redis_conn.set(
+                    task_lock_key,
+                    self.name,
+                    nx=True,
+                    ex=86400
+                )
 
-                        if exclusive_resources:
-                            # This should NOT happen for tasks with exclusive resources!
-                            # We hold the resource locks, so no other worker should be able to claim this task
-                            raise RuntimeError(
-                                f"Worker {self.name} acquired resource locks {exclusive_resources} "
-                                f"for task {task.pk} but another worker claimed it with task lock. "
-                                f"This indicates a serious locking protocol violation."
+                if acquired:
+                    # Successfully claimed the task lock!
+                    # Now try to acquire resource locks for exclusive resources only
+                    # Shared resources are never locked
+                    if exclusive_resources:
+                        if self._try_acquire_resource_locks(exclusive_resources):
+                            # Got both task lock and resource locks!
+                            _logger.info(
+                                "Worker %s acquired task lock and resource locks for task %s in domain: %s",
+                                self.name,
+                                task.pk,
+                                task.pulp_domain.name
                             )
+                            # Store resources so we can release them later
+                            # Note: We don't store the task lock since it auto-expires in 24h
+                            task._locked_resources = exclusive_resources
+                            return task
                         else:
-                            # This is normal for tasks without exclusive resources where multiple workers
-                            # can acquire the (empty) resource locks simultaneously
+                            # Failed to acquire resource locks
+                            # Release the task lock since we can't execute this task
+                            self.redis_conn.delete(task_lock_key)
                             _logger.debug(
-                                "Worker %s tried to claim task %s but another worker got task lock first",
+                                "Worker %s acquired task lock but failed to acquire resource locks for task %s, released task lock",
                                 self.name,
                                 task.pk
                             )
-                            continue
+                            # Add resources to blocked set so we skip other tasks needing them
+                            for resource in exclusive_resources:
+                                blocked_resources.add(resource)
```

Comment:
This parts a little tricky say you have resources [A, B, C] and `_try_acquire_resource_locks()` could acquire A and B, but not C. You want C only to be in this list, but it's going to add A and B. That's not exactly right.

Reply:
I think you should adjust the return interface to `_try_acquire_resource_locks` some to allow the caller to know what could and couldn't be acquired.